packages/sanity/src/core/store/bundles/useBundlesStore.ts

```diff
@@ -0,0 +1,152 @@
+
+  const didInitialFetch = useRef<boolean>(false)
+
+  const initialFetch$ = useCallback(() => {
+    if (!client) {
+      return of(null) // emits null and completes if no client
+    }
+    return client.observable.fetch(QUERY).pipe(
+      timeout(10000), // 10s timeout
+      map((res) => {
+        dispatch({type: 'BUNDLES_SET', bundles: res})
+        didInitialFetch.current = true
+        setLoading(false)
+      }),
+      retry({
+        count: 2,
+        delay: 1000,
+      }),
+      catchError((err) => {
+        if (err.name === 'TimeoutError') {
+          console.error('Fetch operation timed out:', err)
+        }
+        setError(err)
+        return of(null) // ensure stream completion even on error
+      }),
+    )
+  }, [client])
+  const handleListenerEvent = useCallback(
+    (event: ListenEvent<Record<string, BundleDocument>>) => {
+      // Fetch all bundles on initial connection
+      if (event.type === 'welcome' && !didInitialFetch.current) {
+        // Do nothing here, the initial fetch is done in the useEffect below
+        initialFetch$()
+      }
+
+      // The reconnect event means that we are trying to reconnect to the realtime listener.
+      // In this case we set loading to true to indicate that we're trying to
+      // reconnect. Once a connection has been established, the welcome event
+      // will be received and we'll fetch all bundles again (above)
+      if (event.type === 'reconnect') {
+        setLoading(true)
+        didInitialFetch.current = false
+      }
+
+      // Handle mutations (create, update, delete) from the realtime listener
+      // and update the bundles store accordingly
+      if (event.type === 'mutation' && didInitialFetch.current) {
+        if (event.transition === 'disappear') {
+          dispatch({type: 'BUNDLE_DELETED', id: event.documentId})
+        }
+
+        if (event.transition === 'appear') {
+          const nextBundle = event.result as BundleDocument | undefined
+
+          if (nextBundle) {
+            dispatch({type: 'BUNDLE_RECEIVED', payload: nextBundle})
+          }
+        }
+
+        if (event.transition === 'update') {
+          const updatedBundle = event.result as BundleDocument | undefined
+
+          if (updatedBundle) {
+            dispatch({type: 'BUNDLE_UPDATED', payload: updatedBundle})
+          }
+        }
+      }
+    },
+    [initialFetch$],
+  )
+
+  const listener$ = useMemo(() => {
+    if (!client) return of()
+
+    const events$ = client.observable.listen(QUERY, {}, LISTEN_OPTIONS).pipe(
+      map(handleListenerEvent),
+      catchError((err) => {
+        setError(err)
+        return of(err)
+      }),
+    )
+
+    return events$ // as Observable<ListenEvent<Record<string, BundleDocument>>>
+  }, [client, handleListenerEvent])
+
+  useEffect(() => {
+    if (!client) return
+    const subscription = initialFetch$()
+      .pipe(concatMap(() => listener$))
+      .subscribe()
```

Comment:
Instead of manually creating and cleaning up the subscription, would it work to use [the `useObservable` hook from ReactRx](https://react-rx.dev/reference#useobservable)? We use this in some other places already.

Reply:
This is nice, thanks! 
Updated to:
```ts
  const observable = useMemo(() => {
    if (!client) return of(null) // emits null and completes if no client
    return initialFetch$().pipe(concatMap(() => listener$))
  }, [initialFetch$, listener$, client])

  useObservable(observable)
```